pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go

```diff
@@ -28,4 +36,73 @@ type kubeEndpointsController interface {
 	InstancesByPort(c *Controller, svc *model.Service, reqSvcPort int,
 		labelsList labels.Collection) ([]*model.ServiceInstance, error)
 	GetProxyServiceInstances(c *Controller, proxy *model.Proxy, proxyNamespace string) []*model.ServiceInstance
+	updateEDS(obj interface{}, event model.Event)
+}
+
+type kubeEndpoints struct {
+	c        *Controller
+	informer cache.SharedIndexInformer
+}
+
+// GetProxyServiceInstances returns service instances of the given proxy.
+func (e *kubeEndpoints) GetProxyServiceInstances(c *Controller, proxy *model.Proxy, proxyNamespace string) []*model.ServiceInstance {
+	var otherNamespaceEndpoints []*model.ServiceInstance
+	var sameNamespaceEndpoints []*model.ServiceInstance
+
+	for _, item := range e.informer.GetStore().List() {
+		var proxyServiceInstances []*model.ServiceInstance
+		namespace := ""
+		switch item := item.(type) {
+		case *v1.Endpoints:
+			namespace = item.Namespace
+			proxyServiceInstances = getProxyServiceInstancesByEndpoint(c, *item, proxy)
+		case *discoveryv1alpha1.EndpointSlice:
+			namespace = item.Namespace
+			proxyServiceInstances = getProxyServiceInstancesByEndpointSlice(c, item, proxy)
```

Comment:
I mean these code, they still need to know the specific type and call the specific implementation.

Reply:
I see. Made some changes to separate them out. See if the new commit makes sense.